tapiriik/services/PolarFlow/polarflow.py

```diff
@@ -0,0 +1,270 @@
+        return transaction_uri
+
+    def _commit_transaction(self, serviceRecord):
+        if hasattr(serviceRecord, "ServiceData"):
+            transaction_uri = serviceRecord.ServiceData["Transaction-uri"]
+            if transaction_uri:
+                res = requests.put(transaction_uri, headers=self._api_headers(serviceRecord))
+        # TODO : should handle responce code?
+        # 200	OK	Transaction has been committed and data deleted	None
+        # 204	No Content	No content when there is no data available	None
+        # 404	Not Found	No transaction was found with given transaction id	None
+
+    def _api_headers(self, serviceRecord, headers={}):
+        headers.update({"Authorization": "Bearer {}".format(serviceRecord.Authorization["OAuthToken"])})
+        return headers
+
+    def WebInit(self):
+        params = {'response_type':'code',
+                  'client_id': POLAR_CLIENT_ID,
+                  'redirect_uri': WEB_ROOT + reverse("oauth_return", kwargs={"service": "polarflow"})}
+        self.UserAuthorizationURL = "https://flow.polar.com/oauth2/authorization?" + urlencode(params)
+
+    def RetrieveAuthorizationToken(self, req, level):
+        code = req.GET.get("code")
+        params = {"grant_type": "authorization_code",
+                  "code": code,
+                  "redirect_uri": WEB_ROOT + reverse("oauth_return", kwargs={"service": "polarflow"})}
+
+        response = requests.post("https://polarremote.com/v2/oauth2/token", data=params, auth=HTTPBasicAuth(POLAR_CLIENT_ID, POLAR_CLIENT_SECRET))
+        data = response.json()
+
+        if response.status_code != 200:
+            raise APIException(data["error"])
+
+        authorizationData = {"OAuthToken": data["access_token"]}
+        userId = data["x_user_id"]
+
+        try:
+            self._register_user(data["access_token"])
+        except requests.exceptions.HTTPError as err:
+            # Error 409 Conflict means that the user has already been registered for this client.
+            # That error can be ignored
+            if err.response.status_code != 409:
+                raise APIException("Unable to link user", block=True, user_exception=UserException(UserExceptionType.Authorization, intervention_required=True))
+
+        return (userId, authorizationData)
+
+    def RevokeAuthorization(self, serviceRecord):
+        self._delete_user(serviceRecord)
+
+    def SubscribeToPartialSyncTrigger(self, serviceRecord):
+        # There is no per-user webhook subscription with Polar Flow.
+        serviceRecord.SetPartialSyncTriggerSubscriptionState(True)
+
+    def UnsubscribeFromPartialSyncTrigger(self, serviceRecord):
+        # As above.
+        serviceRecord.SetPartialSyncTriggerSubscriptionState(False)
+
+    def PollPartialSyncTrigger(self, multiple_index):
+        response = requests.get(self._api_endpoint + "/v3/notifications", auth=HTTPBasicAuth(POLAR_CLIENT_ID, POLAR_CLIENT_SECRET))
+
+        to_sync_ids = []
```

Comment:
I would make this a set, in case the user has multiple new activities (though I don't think anything goes wrong if it the list has dupes...).

Reply:
Dupes not possible. Result data will contain single entry for each user with a userid, data type and a link to begin-new-transaction url. No matter how much new data is available.

https://www.polar.com/accesslink-api/#list